docs/userguide/strings/cpp-code-point-iterator.md

```diff
@@ -0,0 +1,243 @@
+nav_order: 24
+parent: ICU4C
+---
+<!--
+© 2025 and later: Unicode, Inc. and others.
+License & terms of use: http://www.unicode.org/copyright.html
+-->
+
+# Plug-ins
+{: .no_toc }
+
+## Contents
+{: .no_toc .text-delta }
+
+1. TOC
+{:toc}
+
+---
+
+## Overview
+
+Sometimes you need to process a string one character at a time.
+This is trivial in a UTF-32 string, but those are not common.
+Most Unicode strings are UTF-8 or UTF-16 strings and may use multiple code units per Unicode code point.
+
+(Note that a Unicode code point is not necessarily what you think of as a character.
+See the Wikipedia article on
+[combining characters](https://en.wikipedia.org/wiki/Combining_character)
+for some examples.)
+
+Starting with ICU 78, ICU4C has [C++ header-only APIs](../icu4c/header-only.md)
+for conveniently iterating over the code points of a Unicode string in any standard encoding form (UTF-8/16/32).
+They work seamlessly with modern C++ iterators and ranges.
+These APIs are fully inline-implemented and can be used without linking with the ICU libraries.
+
+As with the existing C macros, there are versions which validate the code unit sequences on the fly,
+as well as fast but “unsafe” versions which assume & require well-formed strings.
+
+Header file documentation:
+[unicode/utfiterator.h](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/utfiterator_8h.html)
+including some
+[sample code snippets](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/utfiterator_8h.html#details).
+
+## Old: C macros
+
+ICU continues to provide C macros for iterating through
+[UTF-8](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/utf8_8h.html)
+and
+[UTF-16](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/utf16_8h.html)
+strings. For example:
+
+```c++
+int32_t rangeLoop16(std::u16string_view s) {
+    // We are just adding up the code points for minimal-code demonstration purposes.
+    const char16_t *p = s.data();
+    size_t length = s.length();
+    int32_t sum = 0;
+    for (size_t i = 0; i < length;) {  // loop body increments i
+        UChar32 c;
+        U16_NEXT(p, i, length, c);
+        sum += c;  // < 0 if ill-formed
+    }
+    return sum;
+}
+```
+
+## C++ code point iterators
+
+The `unicode/utfiterator.h` APIs let you wrap the string in a “range” object that provides iterators over the string’s code points. You could rewrite the C macro example above like this:
+
+```c++
+int32_t rangeLoop16(std::u16string_view s) {
+    // We are just adding up the code points for minimal-code demonstration purposes.
+    int32_t sum = 0;
+    for (auto units : utfStringCodePoints<UChar32, UTF_BEHAVIOR_NEGATIVE>(s)) {
+        sum += units.codePoint();  // < 0 if ill-formed
+    }
+    return sum;
+}
+```
+
+This has a number of benefits compared with the C macros:
+- These C++ APIs provide iterators and range adaptors that are
+  compatible with the C++ standard library, and thus look and feel natural.
```

Comment:
It’s not just that they look and feel natural, it also means they are composable with standard library utilities (especially in C++20 and later).

It might be worth showing one of the ranges-heavy C++20 examples from the test (as you did in your UTW slides), it’s going to be more discoverable here than in the test, and the existing example only really showcases C++11-level constructs.